rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,219 @@
+
+- **switchable** (`any`)
+
+  The feature is fenced by default but a user can opt-in. Preferences may be recorded.
+
+- **disabled** (`off`)
+
+  The feature is fenced and inaccessible, at least not visible. What's ever hidden in open-source?
+
+### Sources
+
+Flags must first be defined somewhere from where they can be loaded into the application.
+
+#### 1. Environment variables
+
+##### Pros
+
+1. Effortless to set up.
+1. Universal way to configure applications.
+1. Low maintenance as every feature can be one env var.
+1. Fairly easy to test considering env vars can be set per-session, per-command, per-container and per-machine (as needed).
+1. Managing a couple of `.env` files is not hard.
+1. I'm not entirely sure but server-side envs could help tree-shake the off features in prod.
+
+##### Cons
+
+1. Will need a deployment to change the environment variables.
+1. Can get complicated if we need to build a lot of scaffolding around it.
+1. Managing a lot of `.env` files is not easy.
+1. Flags will be primitive and will not allow complex data.
+1. Nested flags and flags with metadata will need hacks.
+
+##### Considerations
+
+Feature flag env vars can be prefixed with a string like `FF_` to separate them from the other env vars.
+
+The limitation for the data to be a string means any extra metadata must be JSON-stringified and then parsed in the code. Only primitive flag states can be handled conveniently.
+
+Nesting flags will use hacks like artificial separators e.g. `FF_PARENT__CHILD`. Looks ugly and confusing.
+
+#### 2. Query params
+
+##### Pros
+
+1. Even more effortless to set up.
+1. We've done this before for the embedded mode, [_remember_](https://github.com/WordPress/openverse-frontend/pull/65)?
+1. Testing is easy because tests can append the right flag param to the URL.
+
+##### Cons
+
+1. Configuration will need to live in the codebase so change of state will need deployment.
+1. Exposes half-baked features to users who know the flag.
+1. Transient, so flag must be set on every URL or stored in `localStorage` or cookie.
+1. Flags will be primitive and will not allow complex data. Even stringified JSON would be bad in URLs.
+
+##### Considerations
+
+We can clearly demarcate feature flag params with a prefix such as `ff_` to separate them from the actual query params and process them separately.
+
+Due to their nature, they are best used in pairing with another storage as an override switch rather than on their own.
+
+#### 3. Growthbook
+
+GrowthBook is an open-source platform for feature flagging. It provides SDKs and a UI to manage feature flags.
+
+##### Pros
+
+1. High-level solution with UI.
+1. Supports features we'd want later like A/B testing.
+1. Can record a large number of metrics.
+1. Actively maintained and FOSS.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes.
+
+##### Cons
+
+1. Adds API calls to the server startup.
```

Comment:
I'm not entirely sure (and the [docs](https://nuxtjs.org/docs/concepts/nuxt-lifecycle/#lifecycle) are not doing a great job of clarifying). Essentially Nuxt supports many modes so which steps of the lifecycle run on server start and which run on every request is confusing to me.

Reply:
It would need to be per-request otherwise we would need to restart the server every time the flags changed (for them to get the updated flags). If per-request then maybe with some time based local caching so we avoid calling out on _every_ request and only one request per 15 minute block has to call out to Growthbook.

Reply:
It would need to be per request, prior to Pinia it could have happened in `nuxtServerInit` which is a sort of per-session call. It only runs on the user's first SSR loaded pageview.